File: bookApp/bookApp/bookApp/Services/ChatSocketManager.swift
import Foundation
import Combine

enum SocketConnectionState: Equatable {
    case disconnected
    case connecting
    case connected
    case error(String)
    
    static func == (lhs: SocketConnectionState, rhs: SocketConnectionState) -> Bool {
        switch (lhs, rhs) {
        case (.disconnected, .disconnected),
             (.connecting, .connecting),
             (.connected, .connected):
            return true
        case (.error(let lhsMsg), .error(let rhsMsg)):
            return lhsMsg == rhsMsg
        default:
            return false
        }
    }
}

struct TypingEvent: Codable {
    let userId: String
    let transactionId: String
    let isTyping: Bool
}

struct MessagesReadEvent: Codable {
    let userId: String
    let transactionId: String
}

struct SocketEvent: Codable {
    let event: String
    let data: String?
}

@MainActor
class ChatSocketManager: ObservableObject {
    static let shared = ChatSocketManager()
    
    private var socketURL: String {
        APIConfiguration.shared.socketURL
    }
    
    private var webSocketTask: URLSessionWebSocketTask?
    private var urlSession: URLSession?
    private var currentToken: String?
    private var joinedChats: Set<String> = []
    private var pingTimer: Timer?
    private var reconnectAttempts = 0
    private let maxReconnectAttempts = 10
    
    @Published private(set) var connectionState: SocketConnectionState = .disconnected
    
    let incomingMessages = PassthroughSubject<Message, Never>()
    let typingEvents = PassthroughSubject<TypingEvent, Never>()
    let messagesReadEvents = PassthroughSubject<MessagesReadEvent, Never>()
    
    private init() {}
    
    func connect(token: String) {
        guard connectionState != .connected || currentToken != token else {
            print("[Socket] Already connected with same token")
            return
        }
        
        disconnect()
        currentToken = token
        connectionState = .connecting
        
        guard let url = URL(string: socketURL) else {
            connectionState = .error("Invalid socket URL")
            return
        }
        
        var request = URLRequest(url: url)
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        request.timeoutInterval = 30
        
        let config = URLSessionConfiguration.default
        config.waitsForConnectivity = true
        
        urlSession = URLSession(configuration: config)
        webSocketTask = urlSession?.webSocketTask(with: request)
        
        webSocketTask?.resume()
        
        receiveMessage()
        startPingTimer()
        
        connectionState = .connected
        reconnectAttempts = 0
        print("[Socket] Connected to server")
        
        for chatId in joinedChats {
            joinChat(transactionId: chatId)
        }
    }
    
    func disconnect() {
        pingTimer?.invalidate()
        pingTimer = nil
        
        webSocketTask?.cancel(with: .normalClosure, reason: nil)
        webSocketTask = nil
        urlSession?.invalidateAndCancel()
        urlSession = nil
        
        joinedChats.removeAll()
        connectionState = .disconnected
        print("[Socket] Disconnected from server")
    }
    
    func joinChat(transactionId: String) {
        joinedChats.insert(transactionId)
        
        guard connectionState == .connected else {
            print("[Socket] Cannot join chat - not connected")
            return
        }
        
        let event: [String: Any] = [
            "event": "join_chat",
            "transactionId": transactionId
        ]
        
        sendEvent(event)
        print("[Socket] Joined chat: \(transactionId)")
    }
    
    func leaveChat(transactionId: String) {
        joinedChats.remove(transactionId)
        
        guard connectionState == .connected else { return }
        
        let event: [String: Any] = [
            "event": "leave_chat",
            "transactionId": transactionId
        ]
        
        sendEvent(event)
        print("[Socket] Left chat: \(transactionId)")
    }
    
    func sendTypingStatus(transactionId: String, isTyping: Bool) {
        guard connectionState == .connected else { return }
        
        let event: [String: Any] = [
            "event": "typing",
            "transactionId": transactionId,
            "isTyping": isTyping
        ]
        
        sendEvent(event)
    }
    
    func sendMessageRead(transactionId: String) {
        guard connectionState == .connected else { return }
        
        let event: [String: Any] = [
            "event": "message_read",
            "transactionId": transactionId
        ]
        
        sendEvent(event)
    }
    
    var isConnected: Bool {
        connectionState == .connected
    }
    
    private func sendEvent(_ event: [String: Any]) {
        guard let data = try? JSONSerialization.data(withJSONObject: event),
              let jsonString = String(data: data, encoding: .utf8) else {
            return
        }
        
        let message = URLSessionWebSocketTask.Message.string(jsonString)
        webSocketTask?.send(message) { error in
            if let error = error {
                print("[Socket] Send error: \(error.localizedDescription)")
            }
        }
    }
    
    private func receiveMessage() {
        webSocketTask?.receive { [weak self] result in
            Task { @MainActor in
                switch result {
                case .success(let message):
                    self?.handleMessage(message)
                    self?.receiveMessage()
                    
                case .failure(let error):
                    print("[Socket] Receive error: \(error.localizedDescription)")
                    self?.handleDisconnection()
                }
            }
        }
    }
    
    private func handleMessage(_ message: URLSessionWebSocketTask.Message) {
        switch message {
        case .string(let text):
            parseSocketMessage(text)
        case .data(let data):
            if let text = String(data: data, encoding: .utf8) {
                parseSocketMessage(text)
            }
        @unknown default:
            break
        }
    }
    
    private func parseSocketMessage(_ text: String) {
        guard let data = text.data(using: .utf8),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return
        }
        
        if let eventType = json["0"] as? String {
            handleSocketIOEvent(eventType: eventType, json: json)
        } else if let event = json["event"] as? String {
            handleCustomEvent(event: event, json: json)
        } else {
            handleDirectMessage(json: json)
        }
    }
    
    private func handleSocketIOEvent(eventType: String, json: [String: Any]) {
        switch eventType {
        case "new_message":
            if let messageData = json["1"] as? [String: Any] {
                parseNewMessage(messageData)
            }
        case "message_notification":
            if let notificationData = json["1"] as? [String: Any],
               let messageData = notificationData["message"] as? [String: Any] {
                parseNewMessage(messageData)
            }
        case "user_typing":
            if let typingData = json["1"] as? [String: Any] {
                parseTypingEvent(typingData)
            }
        case "messages_read":
            if let readData = json["1"] as? [String: Any] {
                parseMessagesReadEvent(readData)
            }
        default:
            break
        }
    }
    
    private func handleCustomEvent(event: String, json: [String: Any]) {
        switch event {
        case "new_message":
            if let messageData = json["data"] as? [String: Any] {
                parseNewMessage(messageData)
            }
        case "message_notification":
            if let messageData = json["message"] as? [String: Any] {
                parseNewMessage(messageData)
            }
        case "user_typing":
            parseTypingEvent(json)
        case "messages_read":
            parseMessagesReadEvent(json)
        default:
            break
        }
    }
    
    private func handleDirectMessage(_ json: [String: Any]) {
        if json["id"] != nil && json["content"] != nil && json["senderId"] != nil {
            parseNewMessage(json)
        }
    }
    
    private func parseNewMessage(_ data: [String: Any]) {
        guard let id = data["id"] as? String,
              let content = data["content"] as? String,
              let senderId = data["senderId"] as? String,
              let transactionId = data["transactionId"] as? String,
              let createdAt = data["createdAt"] as? String else {
            return
        }
        
        var sender: MessageSender? = nil
        if let senderData = data["sender"] as? [String: Any],
           let senderId = senderData["id"] as? String,
           let name = senderData["name"] as? String {
            sender = MessageSender(
                id: senderId,
                name: name,
                profileImageUrl: senderData["profileImageUrl"] as? String
            )
        }
        
        let message = Message(
            id: id,
            content: content,
            senderId: senderId,
            transactionId: transactionId,
            createdAt: createdAt,
            isRead: data["isRead"] as? Bool,
            sender: sender
        )
        
        print("[Socket] Received new message: \(id)")
        incomingMessages.send(message)
    }
    
    private func parseTypingEvent(_ data: [String: Any]) {
        guard let userId = data["userId"] as? String,
              let transactionId = data["transactionId"] as? String else {
            return
        }
        
        let event = TypingEvent(
            userId: userId,
            transactionId: transactionId,
            isTyping: data["isTyping"] as? Bool ?? false
        )
        
        typingEvents.send(event)
    }
    
    private func parseMessagesReadEvent(_ data: [String: Any]) {
        guard let userId = data["userId"] as? String,
              let transactionId = data["transactionId"] as? String else {
            return
        }
        
        let event = MessagesReadEvent(
            userId: userId,
            transactionId: transactionId
        )
        
        messagesReadEvents.send(event)
    }
    
    private func startPingTimer() {
        pingTimer?.invalidate()
        pingTimer = Timer.scheduledTimer(withTimeInterval: 25, repeats: true) { [weak self] _ in
            self?.sendPing()
        }
    }
    
    private func sendPing() {
        webSocketTask?.sendPing { [weak self] error in
            if let error = error {
                print("[Socket] Ping failed: \(error.localizedDescription)")
                Task { @MainActor in
                    self?.handleDisconnection()
                }
            }
        }
    }
    
    private func handleDisconnection() {
        connectionState = .disconnected
        pingTimer?.invalidate()
        pingTimer = nil
        
        if reconnectAttempts < maxReconnectAttempts, let token = currentToken {
            reconnectAttempts += 1
            let delay = min(Double(reconnectAttempts) * 1.0, 5.0)
            print("[Socket] Attempting reconnect in \(delay)s (attempt \(reconnectAttempts))")
            
            Task {
                try? await Task.sleep(nanoseconds: UInt64(delay * 1_000_000_000))
                self.connect(token: token)
            }
        } else {
            connectionState = .error("Connection failed after \(maxReconnectAttempts) attempts")
        }
    }
}

File: android-app/app/src/main/java/com/bookstore/bookapp/data/remote/socket/ChatSocketManager.kt
package com.bookstore.bookapp.data.remote.socket

import android.util.Log
import com.bookstore.bookapp.data.remote.api.Message
import com.bookstore.bookapp.data.remote.api.MessageSender
import com.google.gson.Gson
import com.google.gson.JsonObject
import io.socket.client.IO
import io.socket.client.Socket
import io.socket.emitter.Emitter
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.callbackFlow
import org.json.JSONObject
import java.net.URI
import com.bookstore.bookapp.di.NetworkModule
import javax.inject.Inject
import javax.inject.Singleton

sealed class SocketConnectionState {
    data object Disconnected : SocketConnectionState()
    data object Connecting : SocketConnectionState()
    data object Connected : SocketConnectionState()
    data class Error(val message: String) : SocketConnectionState()
}

data class TypingEvent(
    val userId: String,
    val transactionId: String,
    val isTyping: Boolean
)

data class MessagesReadEvent(
    val userId: String,
    val transactionId: String
)

@Singleton
class ChatSocketManager @Inject constructor() {
    
    companion object {
        private const val TAG = "ChatSocketManager"
        private val SOCKET_URL = NetworkModule.SOCKET_URL
    }
    
    private var socket: Socket? = null
    private val gson = Gson()
    private var currentToken: String? = null
    private val joinedChats = mutableSetOf<String>()
    
    private val _connectionState = MutableStateFlow<SocketConnectionState>(SocketConnectionState.Disconnected)
    val connectionState: StateFlow<SocketConnectionState> = _connectionState.asStateFlow()
    
    private val _incomingMessages = MutableSharedFlow<Message>(replay = 0, extraBufferCapacity = 64)
    val incomingMessages: SharedFlow<Message> = _incomingMessages.asSharedFlow()
    
    private val _typingEvents = MutableSharedFlow<TypingEvent>(replay = 0, extraBufferCapacity = 16)
    val typingEvents: SharedFlow<TypingEvent> = _typingEvents.asSharedFlow()
    
    private val _messagesReadEvents = MutableSharedFlow<MessagesReadEvent>(replay = 0, extraBufferCapacity = 16)
    val messagesReadEvents: SharedFlow<MessagesReadEvent> = _messagesReadEvents.asSharedFlow()
    
    fun connect(token: String) {
        if (socket?.connected() == true && currentToken == token) {
            Log.d(TAG, "Already connected with same token")
            return
        }
        
        disconnect()
        currentToken = token
        _connectionState.value = SocketConnectionState.Connecting
        
        try {
            val options = IO.Options().apply {
                auth = mapOf("token" to token)
                transports = arrayOf("websocket", "polling")
                reconnection = true
                reconnectionAttempts = 10
                reconnectionDelay = 1000
                reconnectionDelayMax = 5000
                timeout = 20000
            }
            
            socket = IO.socket(URI.create(SOCKET_URL), options).apply {
                on(Socket.EVENT_CONNECT, onConnect)
                on(Socket.EVENT_DISCONNECT, onDisconnect)
                on(Socket.EVENT_CONNECT_ERROR, onConnectError)
                on("new_message", onNewMessage)
                on("message_notification", onMessageNotification)
                on("user_typing", onUserTyping)
                on("messages_read", onMessagesRead)
                connect()
            }
            
            Log.d(TAG, "Attempting to connect to socket server")
        } catch (e: Exception) {
            Log.e(TAG, "Failed to create socket connection", e)
            _connectionState.value = SocketConnectionState.Error(e.message ?: "Connection failed")
        }
    }
    
    fun disconnect() {
        socket?.let { s ->
            s.off()
            s.disconnect()
            Log.d(TAG, "Disconnected from socket server")
        }
        socket = null
        joinedChats.clear()
        _connectionState.value = SocketConnectionState.Disconnected
    }
    
    fun joinChat(transactionId: String) {
        if (joinedChats.contains(transactionId)) {
            Log.d(TAG, "Already joined chat: $transactionId")
            return
        }
        
        socket?.let { s ->
            if (s.connected()) {
                s.emit("join_chat", transactionId)
                joinedChats.add(transactionId)
                Log.d(TAG, "Joined chat: $transactionId")
            } else {
                Log.w(TAG, "Cannot join chat - socket not connected")
            }
        }
    }
    
    fun leaveChat(transactionId: String) {
        socket?.let { s ->
            if (s.connected()) {
                s.emit("leave_chat", transactionId)
                joinedChats.remove(transactionId)
                Log.d(TAG, "Left chat: $transactionId")
            }
        }
    }
    
    fun sendTypingStatus(transactionId: String, isTyping: Boolean) {
        socket?.let { s ->
            if (s.connected()) {
                val data = JSONObject().apply {
                    put("transactionId", transactionId)
                    put("isTyping", isTyping)
                }
                s.emit("typing", data)
            }
        }
    }
    
    fun sendMessageRead(transactionId: String) {
        socket?.let { s ->
            if (s.connected()) {
                val data = JSONObject().apply {
                    put("transactionId", transactionId)
                }
                s.emit("message_read", data)
            }
        }
    }
    
    fun isConnected(): Boolean = socket?.connected() == true
    
    private val onConnect = Emitter.Listener {
        Log.d(TAG, "Socket connected")
        _connectionState.value = SocketConnectionState.Connected
        
        joinedChats.forEach { chatId ->
            socket?.emit("join_chat", chatId)
        }
    }
    
    private val onDisconnect = Emitter.Listener { args ->
        val reason = args.firstOrNull()?.toString() ?: "Unknown"
        Log.d(TAG, "Socket disconnected: $reason")
        _connectionState.value = SocketConnectionState.Disconnected
    }
    
    private val onConnectError = Emitter.Listener { args ->
        val error = args.firstOrNull()?.toString() ?: "Unknown error"
        Log.e(TAG, "Socket connection error: $error")
        _connectionState.value = SocketConnectionState.Error(error)
    }
    
    private val onNewMessage = Emitter.Listener { args ->
        try {
            val data = args.firstOrNull()
            if (data != null) {
                val message = parseMessage(data)
                message?.let {
                    Log.d(TAG, "Received new message: ${it.id}")
                    _incomingMessages.tryEmit(it)
                }
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error parsing new message", e)
        }
    }
    
    private val onMessageNotification = Emitter.Listener { args ->
        try {
            val data = args.firstOrNull()
            if (data != null && data is JSONObject) {
                val messageJson = data.optJSONObject("message")
                messageJson?.let {
                    val message = parseMessage(it)
                    message?.let { msg ->
                        Log.d(TAG, "Received message notification: ${msg.id}")
                        _incomingMessages.tryEmit(msg)
                    }
                }
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error parsing message notification", e)
        }
    }
    
    private val onUserTyping = Emitter.Listener { args ->
        try {
            val data = args.firstOrNull() as? JSONObject
            data?.let {
                val event = TypingEvent(
                    userId = it.optString("userId"),
                    transactionId = it.optString("transactionId"),
                    isTyping = it.optBoolean("isTyping", false)
                )
                _typingEvents.tryEmit(event)
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error parsing typing event", e)
        }
    }
    
    private val onMessagesRead = Emitter.Listener { args ->
        try {
            val data = args.firstOrNull() as? JSONObject
            data?.let {
                val event = MessagesReadEvent(
                    userId = it.optString("userId"),
                    transactionId = it.optString("transactionId")
                )
                _messagesReadEvents.tryEmit(event)
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error parsing messages read event", e)
        }
    }
    
    private fun parseMessage(data: Any): Message? {
        return try {
            val jsonStr = when (data) {
                is JSONObject -> data.toString()
                is String -> data
                else -> return null
            }
            
            val json = JSONObject(jsonStr)
            val senderJson = json.optJSONObject("sender")
            
            Message(
                id = json.optString("id"),
                transactionId = json.optString("transactionId"),
                senderId = json.optString("senderId"),
                content = json.optString("content"),
                isRead = json.optBoolean("isRead", false),
                createdAt = json.optString("createdAt"),
                sender = if (senderJson != null) {
                    MessageSender(
                        id = senderJson.optString("id"),
                        name = senderJson.optString("name"),
                        profileImageUrl = senderJson.optString("profileImageUrl").takeIf { it.isNotEmpty() }
                    )
                } else {
                    MessageSender(id = "", name = "Unknown", profileImageUrl = null)
                }
            )
        } catch (e: Exception) {
            Log.e(TAG, "Error parsing message JSON", e)
            null
        }
    }
}
